[DI] Allow fetching env vars with lookup-dedicated services

nicolas-grekas opened the pull request:
| Q | A |
| --- | --- |
| Branch? | 3.4 |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | (not written yet) |
| Fixed tickets | - |
| License | MIT |
| Doc PR | (if feat. accepted) |

Thinking about issues like e.g. https://github.com/platformsh/platformsh-example-symfony/issues/16, that need some preprocessing to get env vars values, one can already create a custom base container class and override the `getEnv()` protected method to achieve it.
But unless I missed something, one couldn't write a bundle to package this for easy reuse.
I propose to ship 3.2 with the attached patch, that adds a way to reference env var using this syntax:
`%env(var@service)%`, that would turn into a dynamic call to `$container->get('service')->getEnv('var')` (well, not exactly so that it works with private services also, but that's the spirit).
Before writing any tests and fine tuning the feature, WDYT of the idea?

nicolas-grekas committed:
[DI] Allow fetching env vars with lookup-dedicated services

ro0NL commented:
What about a function-like syntax `%env(VAR[, lookup_service])%`?

javiereguiluz commented:
I can't propose a better alternative, but the proposed syntax looks ugly to me. Besides, Symfony 3.2 entered the feature freeze period a month ago, so I'm not sure if it's a good idea to consider this new feature for 3.2.

nicolas-grekas commented:
@javiereguiluz I'd love to have a better proposal! The feat. freeze perdiod is exactly for this: fine tuning new features (and this feat. may qualify as such), and tweaking their API.
For ex., if you think that `%env(FOO)%` should be `%env$FOO%`, and `%env(service:FOO)%` be `%env$service$FOO%`, it's now, of never :)

javiereguiluz commented:
After thinking a bit about this, I like the syntax proposed by @ro0NL:

```yaml
# normal env variable
%env(DB_PASSWORD)%

# env variable with a lookup service
%env(DB_PASSWORD, @app.lookup_service)%
```

nicolas-grekas commented:
You added a space, then a comma: this is the beginning of a DSL... I understand it looks sexy, but that's overkill to me here. I prefer seeing this as a namespaced identifier.

ro0NL commented:
Btw, i proposed it like this because the current syntax already was like `something(value)`, looks like a function to me ;-) That was the beginning of a DSL imo.

However agree that this syntax may not be desired given the context; represent a value placeholder.

What about array-like syntax then?
```
%env[FOO]%
%env:lookup_service[FOO]%
```

ro0NL commented:
And last but not least, we should consider to differ between a _lookup_ and a _decorating_ service.

```yml
# lookup service
%env[key]%
%lookup_service[key]%

# decorating service
%env:decorating_service[key]
%lookup_service:decorating_service[key]%
```

Decorating services receive the key as well as the value (from the lookup service),.

chalasr commented on src/Symfony/Component/DependencyInjection/Container.php:
```diff
@@ -419,6 +419,14 @@ protected function getEnv($name)
         if (false !== $env = getenv($name)) {
             return $this->envCache[$name] = $env;
         }
+        if (0 < $i = strpos($name, ':')) {
```
Shouldn't this check be the first?

nicolas-grekas replied:
It could. I made it this way on purpose though, looks more flexible to me.

chalasr commented on src/Symfony/Component/DependencyInjection/Container.php:
```diff
@@ -419,6 +419,14 @@ protected function getEnv($name)
+            $id = substr($name, 0, $i);
+            $service = isset($this->services[$id]) ? $this->services[$id] : (isset($this->methodMap[$id]) ? $this->{$this->methodMap[$id]}() : $this->get($id));
+
+            if (null !== $env = $service->getEnv(substr($name, 1 + $i))) {
```
Don't we need an interface for services with such ability, ensuring the `getEnv()` method?

ro0NL replied:
With my last example, i propose to have a `ParameterLookupInterface::lookup` and `ParamaterDecoratorInterface::decorate`.

nicolas-grekas replied:
Adding the interface was next on my list. So is a compiler pass that checks that the referenced services exist and implement it.

ro0NL replied:
What about `processEnv($name, $value)` then? And continue with the `%env:service(key)%` syntax (or square brackets, i dont mind). We should avoid the term _lookup_.

Besides you can still process `$value` based on `$key` if needed (ie. do a lookup). But i think simply decorating `$value` is most common.

nicolas-grekas replied:
interface & compiler pass added

nicolas-grekas commented:
> If we go one step further it could be %<service_id>[<key_to_get>]%;

We have to remember that the env-related syntax is potentially conflicting with existing parameters into the wild. Having such a generic syntax makes the collision much more likely to me.

> decorating service

I don't see the need for these. Lookup services are required for operations such as json-decoding some env var. There is no point into architecturing for greater extensibility here IMHO.

ro0NL commented:
> Having such a generic syntax makes the collision much more likely to me.

We could always check those first right? But im fine with a `env` prefix as well because of `EnvPlaceholderParameterBag`.

Just saying if we allow to change the underlying implementation it might _not_ be env related anymore, ie. `%env:lookup_from_db[key]%`, which perhaps makes less sense then having `%db[key]%`.

> decorating service

I thought of it in terms of letting symfony lookup the value from the env, and my application decorate it. E.g. do things as `string("true") > bool(true)`.

nicolas-grekas commented:
I'm giving up on the idea to merge this in 3.2 :)
Syntax updated to ``%env(FOO_MANAGED_ENV_VAR@foo_lookup_service)%`.

> it might not be env related anymore

true, but should we care? env-like is enough. Your example (db) would be considered a bad practice: env lookups should be blazing-fast, and available early at bootstrap (because they are for configuring services, such as eg. ... db connections :) )

> things as `string("true")` > `bool(true)`.

So, totally unrelated to this PR :) And I fail to see what issue this would solve btw (but please discuss this is a dedicated issue then if you have one).

ro0NL commented:
> Lookup services are required for operations such as json-decoding some env var.

Sound like decorating to me ;-) assuming symfony still provides the env var.

But i get your point.. the related issue https://github.com/platformsh/platformsh-example-symfony/issues/16#issuecomment-248900141 is actually lookup related;

>  reading from a base64+json-encoded env var

ro0NL commented on src/Symfony/Component/DependencyInjection/Container.php:
```diff
@@ -419,6 +419,14 @@ protected function getEnv($name)
         if (false !== $env = getenv($name)) {
             return $this->envCache[$name] = $env;
         }
+        if (false !== $i = strpos($name, '@')) {
```
What about a default implementation of `GetEnvInterface`? and be explicit here (in terms of priority);

```php
if () {
   $service->getEnv();
} else {
   $default->getEnv();
}
```

nicolas-grekas replied:
We already have a default here: it's the code surrounding the new block.

ro0NL replied:
I mean a `GetEnv` class for the `getenv/$_ENV` lookup.

nicolas-grekas replied:
Yep, that's what I thought, and that would require a class, which means a service, thus a service definition. But the container comes without any service preconfigured...

ro0NL replied:
This is about if we should allow to override the lookup service by having `FOO@service` in our `$_ENV` var for example. I've no strong opinion on this, but it could make sense to do:

```php
if (false !== $i = strpos($name, '@')) {
    $value = $service->getEnv(substr($name, 0, $i));
} else {
   $default = new GetEnv();
   $value = $default->getEnv($name);
}
```

ro0NL replied:
Oh and shouldnt this be `strrpos`? Not sure ;-)

ro0NL commented on src/Symfony/Component/DependencyInjection/Container.php:
```diff
@@ -419,6 +419,14 @@ protected function getEnv($name)
+                return $this->envCache[$name] = $env;
+            }
+        }
         if (!$this->hasParameter("env($name)")) {
```
Shouldnt we use the normalized name here? And provide a default regarding the lookup service?

nicolas-grekas replied:
By normalized name, you mean with lowercased service id? If yes, it's not required: parameters are already case insensitive in the base ParameterBag.
And no, we don't want a default service: the container comes with *no* service by default, yet env vars should work. See comment above.

nicolas-grekas replied:
But these are really two separate variable identifiers.

ro0NL reviewed (commented):
> I'm giving up the idea to merge this in 3.2.

Dont :) This is good. Syntax is fine as well imo.

nicolas-grekas commented:
Tests added

chalasr reviewed (commented):
LGTM 👍

javiereguiluz commented:
Regarding the custom syntax ... let's try to make it compatible with the idea of adding new features in the future. For example, in #20434 they are asking us for a way to transform env vars to other types different than strings (e.g. bool, int, float).

If we decide to use the same syntax as the ParameterBag, we could allow for things like:

```
%env(getBoolean(SF_DEBUG))%
%env(getInt(SF_MAX_CONNECTIONS))%
```

nicolas-grekas commented:
This is not a DSL, but a namespaced string.
So could be `int:FOO`.
BUT let's not plan for something that might never happen. #20434 is far from being ready nor easy nor confirmed...

nicolas-grekas commented:
Closing this one, because it may not be interesting enough - or it may be better for people to align on how env vars works.
Could be resurrected if someone else comes with a use case he actually needs.

ElectricMaxxx committed:
restrict reflection doc block

The version 3.2.0 and 3.2.1 of reflection-docblock is broken and lower version as 3.1 miss some tags

nicolas-grekas committed:
[DI] Fix YamlDumper not dumping abstract and autoconfigure

nicolas-grekas committed:
bug #23854 [DI] Fix YamlDumper not dumping abstract and autoconfigure (nicolas-grekas)

This PR was merged into the 3.3 branch.

Discussion
----------

[DI] Fix YamlDumper not dumping abstract and autoconfigure

| Q             | A
| ------------- | ---
| Branch?       | 3.3
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | -
| License       | MIT
| Doc PR        | -

Commits
-------

685ff0e [DI] Fix YamlDumper not dumping abstract and autoconfigure

nicolas-grekas committed:
[DI] Fix dumping abstract with YamlDumper

nicolas-grekas committed:
bug #23848 restrict reflection doc block (ElectricMaxxx)

This PR was merged into the 3.3 branch.

Discussion
----------

restrict reflection doc block

| Q             | A
| ------------- | ---
| Branch?       | 3.1, 3.2, 3.3, master
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| License       | MIT

I think we should have same behavior on 2.8 and 3.0 but there `phpdocumentor/reflection` is required and i whanted to avoid conflicts.

### Reason:

The version 3.2.0 and 3.2.1 of phpdocumentor/reflection-docblock is really broken. All Annotations lide `@api`, means without any bracket, leads to errors like `Uninitialized string offset: 0`. We in the CMF do not require that bundle directly, but symfony does and so we get braking builds.

Commits
-------

6760127 restrict reflection doc block

nicolas-grekas committed:
bug #23856 [DI] Fix dumping abstract with YamlDumper (nicolas-grekas)

This PR was merged into the 2.7 branch.

Discussion
----------

[DI] Fix dumping abstract with YamlDumper

| Q             | A
| ------------- | ---
| Branch?       | 2.7
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | -
| License       | MIT
| Doc PR        | -

Commits
-------

c396e8c [DI] Fix dumping abstract with YamlDumper

nicolas-grekas committed:
Merge branch '2.7' into 2.8

* 2.7:
  [DI] Fix dumping abstract with YamlDumper

nicolas-grekas committed:
Merge branch '2.8' into 3.3

* 2.8:
  [DI] Fix dumping abstract with YamlDumper

nicolas-grekas committed:
Merge branch '3.3' into 3.4

* 3.3:
  [DI] Fix dumping abstract with YamlDumper
  restrict reflection doc block
  [DI] Fix YamlDumper not dumping abstract and autoconfigure

nicolas-grekas committed:
fix merge
